File: go/nrpc/nrpc_client.go
/*
 * @Author: FengyunSky lifengsky123@gmail.com
 * @Description: 对接核心 rpc 层实现远程过程调用，实现核心的rpc主要调用API
 */

package nrpc

import (
	"fmt"
	"strings"
	"time"

	"github.com/nats-io/nats.go"
	"google.golang.org/protobuf/proto"
)

const (
	LogTagClient = "nrpclient"
)

type QosType int8

// 定义消息的qos
const (
	Qos0 QosType = 0 //不保证消息不丢失
	Qos1 QosType = 1 //保证消息不丢失，但不保证消息唯一
	Qos2 QosType = 2 //保证消息不丢失，且保证消息唯一
)

type RequestOpt struct {
	Qos     QosType       //消息的qos
	MsgId   string        //消息唯一ID(只有qos2需要指定，其他情况丢弃)
	Timeout time.Duration //消息的过期时间
}

func RequestDefaultOpt() *RequestOpt {
	return &RequestOpt{
		Qos:     Qos0,
		MsgId:   "",
		Timeout: 10 * time.Second,
	}
}

type NRPCClient struct {
	NRpc *NRPC
}

func NewNRPCClient(name, subject, encoding string, logger Logger, nc NatsClient) *NRPCClient {
	CheckEncode(encoding)
	CheckSubject(subject)
	return &NRPCClient{
		NRpc: NewNPRC(name, subject, encoding, nc, logger),
	}
}

/**
 * @name: CallSync
 * @msg: 同步调用
 * @param {proto.Message} req
 * @param {proto.Message} rsp
 * @param {*RequestOpt} opt
 * @return {error}
 */
func (client *NRPCClient) CallSync(req proto.Message, rsp proto.Message, opt *RequestOpt) error {
	// encode request
	rawRequest, err := Marshal(client.NRpc.encoding, req)
	if err != nil {
		client.NRpc.logger.Errorf(LogTagClient, "Marshal err: %v", err)
		return fmt.Errorf("CallSync failed, Marshal err: %v", err)
	}

	subject := client.NRpc.subject + "." + strings.ToLower(GetCallerFuncName())
	stream, _ := client.checkRequestOpt(opt, subject)
	client.NRpc.logger.Infof(LogTagClient, "handler the subject:%s qos:%d timeout:%d msgid:%s",
		subject, opt.Qos, opt.Timeout/time.Millisecond, opt.MsgId)

	if opt.Qos == Qos0 {
		reply, err := client.NRpc.nc.Request(subject, rawRequest, opt.Timeout)
		if err != nil {
			client.NRpc.logger.Errorf(LogTagClient, "Request err: %v", err)
			return fmt.Errorf("Request err: %v", err)
		}

		if err := UnmarshalResponse(client.NRpc.encoding, reply.Data, rsp); err != nil {
			client.NRpc.logger.Errorf(LogTagClient, "response unmarshal err:%v", err)
			return fmt.Errorf("CallSync failed, UnmarshalResponse err: %v", err)
		}
	} else {
		replyChan, err := client.request(subject, stream, rawRequest, opt)
		if err != nil {
			client.NRpc.logger.Errorf(LogTagClient, "request err:%v", err)
			return fmt.Errorf("Call failed, request err: %v", err)
		}

		select {
		case msg := <-replyChan:
			client.NRpc.logger.Infof(LogTagClient, "response for subject:%s len:%d", msg.Subject, len(msg.Data))
			if err := UnmarshalResponse(client.NRpc.encoding, msg.Data, rsp); err != nil {
				client.NRpc.logger.Errorf(LogTagClient, "response for subject:%s unmarshal err:%v", msg.Subject, err)
				return fmt.Errorf("CallSync failed, response for subject:%s unmarshal err:%v", msg.Subject, err)
			}
		case <-time.After(opt.Timeout):
			client.NRpc.logger.Errorf(LogTagClient, "reply timeout")
			return fmt.Errorf("CallSync failed, err: timeout")
		}
	}

	return nil
}

type ResponseCallback func(*Response, error)

/**
 * @name: Call
 * @msg: 异步调用
 * @param {proto.Message} req
 * @param {ResponseCallback} cb
 * @param {*RequestOpt} opt
 * @return {error}
 */
func (client *NRPCClient) Call(req proto.Message, cb ResponseCallback, opt *RequestOpt) error {
	// encode request
	rawRequest, err := Marshal(client.NRpc.encoding, req)
	if err != nil {
		client.NRpc.logger.Errorf(LogTagClient, "Marshal err: %v", err)
		return fmt.Errorf("Call failed, Marshal err: %v", err)
	}

	reply := nats.NewInbox()
	replyChan := make(chan *nats.Msg)
	//服务名称转化为小写，避免不同语言的命名差异
	subject := client.NRpc.subject + "." + strings.ToLower(GetCallerFuncName())

	client.NRpc.logger.Infof(LogTagClient, "handler the subject:%s qos:%d timeout:%d msgid:%s",
		subject, opt.Qos, opt.Timeout/time.Millisecond, opt.MsgId)

	//检查opt是否合法并修改为合法的opt
	stream, _ := client.checkRequestOpt(opt, subject)
	if opt.Qos == Qos0 {
		_, err := client.NRpc.nc.ChanSubscribe(reply, replyChan)
		if err != nil {
			client.NRpc.logger.Errorf(LogTagClient, "ChanSubscribe err:%v", err)
			return fmt.Errorf("Call failed, ChanSubscribe err: %v", err)
		}
		if err := client.NRpc.nc.PublishRequest(subject, reply, rawRequest); err != nil {
			client.NRpc.logger.Errorf(LogTagClient, "PublishRequest err: %v", err)
			return fmt.Errorf("Call failed, PublishRequest err: %v", err)
		}
	} else {
		replyChan, err = client.request(subject, stream, rawRequest, opt)
		if err != nil {
			client.NRpc.logger.Errorf(LogTagClient, "request err:%v", err)
			return fmt.Errorf("Call failed, request err: %v", err)
		}
	}

	go func() {
		for {
			select {
			case msg := <-replyChan:
				client.NRpc.logger.Errorf(LogTagClient, "reply msg for subject:%s reply:%s len:%d",
					msg.Subject, msg.Reply, len(msg.Data))
				//若推送消息为stream会存在PubAck，这里需要去除
				data := msg.Data
				reply := string(data)
				if strings.Contains(reply, "\"stream\"") || strings.Contains(reply, "+ACK") ||
					strings.Contains(reply, "+WPI") {
					client.NRpc.logger.Errorf(LogTagClient,
						"reply for subject %s is the PubAck message or +ACK, maybe the stream message, to drop",
						msg.Subject)
					continue
				}

				var rep Response
				if err := UnmarshalResponse(client.NRpc.encoding, data, &rep); err != nil {
					client.NRpc.logger.Errorf(LogTagClient, "response unmarshal err:%v", err)
					cb(nil, err)
					return
				}
				cb(&rep, nil)
				return
			case <-time.After(opt.Timeout):
				cb(nil, nats.ErrTimeout)
			}
		}
	}()
	return nil
}

/**
 * @name: Notify
 * @msg: 广播消息不接收处理响应
 * @param {proto.Message} req
 * @param {string} subject
 * @return {error}
 */
func (client *NRPCClient) Notify(req proto.Message, subject string) error {
	// encode request
	rawRequest, err := Marshal(client.NRpc.encoding, req)
	if err != nil {
		client.NRpc.logger.Errorf(LogTagClient, "Marshal err: %v", err)
		return fmt.Errorf("Notify failed, Marshal err: %v", err)
	}

	subject = subject + "." + GetCallerFuncName()
	err = client.NRpc.nc.Publish(subject, rawRequest)
	if err != nil {
		client.NRpc.logger.Errorf(LogTagClient, "Notify failed, Publish err: %v", err)
		return fmt.Errorf("Notify failed, Publish err: %v", err)
	}

	return nil
}

func (client *NRPCClient) request(subject, stream string, data []byte, opt *RequestOpt) (chan *nats.Msg, error) {
	// 定义响应回复的主题为添加后缀.reply 来处理
	replyChan := make(chan *nats.Msg)
	reply := nats.NewInbox() + "." + subject
	client.NRpc.logger.Infof(LogTagServer, "request for subject %s by %s", subject, reply)
	_, err := client.NRpc.nc.ChanSubscribe(reply, replyChan)
	if err != nil {
		client.NRpc.logger.Errorf(LogTagClient, "ChanSubscribe err:%v", err)
		return replyChan, fmt.Errorf("Request failed, ChanSubscribe err: %v", err)
	}

	client.NRpc.logger.Infof(LogTagServer, "publish the msg for the subject:%s", subject)
	ackWait := opt.Timeout / 2
	opts := []nats.PubOpt{
		nats.ExpectStream(stream),
		nats.AckWait(ackWait),
		nats.RetryWait(ackWait / 3),
		nats.RetryAttempts(3),
	}
	if opt.Qos == Qos2 && opt.MsgId != "" {
		opts = append(opts, nats.MsgId(opt.MsgId))
	}

	err = client.NRpc.nc.PublishMsg(&nats.Msg{
		Subject: subject,
		Header: nats.Header{
			ExpectedReplySujectHdr: []string{reply}, //通过header通过响应主题
		},
		Data: data},
		opts...)

	if err != nil {
		client.NRpc.logger.Errorf(LogTagClient, "pushlish err: %v", err)
		return replyChan, fmt.Errorf("Request failed, pushlish err: %v", err)
	}

	return replyChan, nil
}

/**
 * @name: checkRequestOpt
 * @msg: 校验请求选项并修正qos
 * @param {*RequestOpt} opt
 * @param {string} subject
 * @return {string, error}
 */
func (client *NRPCClient) checkRequestOpt(opt *RequestOpt, subject string) (string, error) {
	if opt == nil {
		opt = RequestDefaultOpt()
	}

	if opt.Timeout == 0 {
		client.NRpc.logger.Panicf(LogTagClient, "invalid timeou(0)")
	}

	stream, err := client.NRpc.nc.StreamNameBySubject(subject)
	//判定订阅主题是否存在stream，若订阅主题包含了stream则需要使用Qos1/2
	if opt.Qos == Qos0 {
		if err != nil {
			client.NRpc.logger.Warnf(LogTagClient,
				"lookup the stream for subject %s, by StreamNameBySubject err:%v",
				subject, err)
			return "", err
		} else {
			if opt.MsgId != "" {
				opt.Qos = Qos2
			} else {
				opt.Qos = Qos1
			}
		}
	}

	return stream, nil
}

File: js/nrpc/nrpc_client.js
/*
 * @Author: FengyunSky lifengsky123@gmail.com
 * @Description: 对接核心 rpc 层实现远程过程调用，实现核心的rpc主要调用API
 */

const Assert = require('assert');
const Is = require('is');
const Util = require('util');
const Nats = require('nats');
const {NRPC, Request, Response} = require('./nrpc');
const {EncodeType, EncodeType1, ExpectedReplySujectHdr} = require('./nrpc');

const Qos = {
    Qos0: 0,//不保证消息不丢失
    Qos1: 1,//保证消息不丢失，但不保证消息唯一
    Qos2: 2,//保证消息不丢失，且保证消息唯一
}
class RequestOpt {
    constructor(qos = Qos.Qos0, msgID = '', timeout = 10*1000) {
        Assert(qos >= Qos.Qos0 && qos <= Qos.Qos2, 'RequestOpt failed, invalid qos');
        Assert(Is.string(msgID), 'RequestOpt failed, invalid msgid');
        Assert(Is.number(timeout), 'RequestOpt failed, invalid timeout');
        this.qos = qos; //消息的qos
        this.msgID = msgID;//消息唯一ID(只有qos2需要指定，其他情况丢弃)
        this.timeout = timeout;//消息的过期时间
    }

    getQos() {
        return this.qos;
    }

    setQos(qos) {
        Assert(qos >= Qos.Qos0 && qos <= Qos.Qos2, 'RequestOpt failed, invalid qos');
        this.qos = qos;
    }

    getMsgID() {
        return this.msgID;
    }

    getTimeout() {
        return this.timeout;
    }
}

class NRPCClient extends NRPC {
    constructor(name, subject, client, encodeType) {
        super(name, subject, client, encodeType);
    }

    /**
     * @name: callSync
     * @msg: 同步调用
     * @param {Request} request
     * @param {RequestOpt} opts
     * @return {Promise(Response)}
     */    
    async callSync(request, opts) {
        Assert(request, 'callSync failed, invalid request');
        Assert(opts instanceof RequestOpt, 'callSync failed, invalid opts');
        let req = NRPC.encode(this.encodeType, request, EncodeType1.Request);
        let subject = this.subject + '.' + NRPC.getCallerFuncName().toLowerCase();
        console.log(Util.format('callSync, handler the subject:%s, qos:%d timeout:%d msgid:%s', 
                        subject, opts.getQos(), opts.getTimeout()/1000, opts.getMsgID()));

        if (opts.getQos() === Qos.Qos0) {
            try {
                let msg = await this.client.request(subject, req, opts.getTimeout());
                let err = Response.verify(msg.data);
                if (err) {
                    return Promise.reject(new Error(Util.format('callSync failed, response verify err:%s', err)));
                }
                let reponse = NRPC.decode(this.encodeType, msg.data);
                return Promise.resolve(reponse);
            } catch (err) {
                return Promise.reject(new Error(Util.format('callSync failed, request err:%s', err.message)));
            }
        } else {
            let time = setTimeout(function(){
                return Promise.reject(new Error('callSync failed, err: timeout'));
            }, opts.getTimeout());

            try {
                let stream = await this._checkRequestOpt(opts, subject);
                let msg = await this._request(subject, stream, req, opts);

                clearTimeout(time);
                let err = Response.verify(msg.data);
                if (err) {
                    return Promise.reject(new Error(Util.format('callSync failed, response verify err:%s', err)));
                }
                let reponse = NRPC.decode(this.encodeType, msg.data);
                return Promise.resolve(reponse);
            } catch (err) {
                clearTimeout(time);
                return Promise.reject(new Error(Util.format('callSync failed, request err:%s', err.message)));
            }
            
        }
    }

    /**
     * @name: call
     * @msg: 异步调用
     * @param {string|Uint8Array} request
     * @param {RequestOpt} opts
     * @param {function(Error, Response)} callback
     * @return {Promise}
     */    
    async call(request, opts, callback) {
        Assert(request, 'call failed, invalid request');
        Assert(Is.function(callback), 'call failed, invalid callback');
        Assert(opts instanceof RequestOpt, 'call failed, invalid opts');

        let req = NRPC.encode(this.encodeType, request, EncodeType1.Request);
        let subject = this.subject + '.' + request.name.toLowerCase();
        console.log(Util.format('call, handler the subject:%s, qos:%d timeout:%d msgid:%s', 
                        subject, opts.getQos(), opts.getTimeout()/1000, opts.getMsgID()));

        let time = setTimeout(function(){
            callback(new Error('call failed, err: timeout'), null);
        }, opts.getTimeout());

        let that = this;
        let handler = function(err, msg, callback) {
            if (err) {
                console.log(Util.format('[call]recv subject:%s failed, err:%s', reply, err.message));
                return;
            }
            let errMsg = Response.verify(msg.data);
            if (errMsg) {
                callback(new Error(Util.format('call failed, response verify err:%s', errMsg)), null);
                return;
            }
            let reponse = NRPC.decode(that.encodeType, msg.data);
            callback(null, reponse);
        };
        
        if (opts.getQos() === Qos.Qos0) {
            try {
                let reply = Nats.createInbox() + '.' + subject;
                await this.client.subscribe(reply, function(err, msg){
                    clearTimeout(time);
                    handler(err, msg, callback);
                })
                this.client.publishRequest(subject, reply, req);
            } catch (err) {
                return Promise.reject(new Error(Util.format('call failed, subscribe err:%s', err.message)));
            }
        } else {
            let stream = await this._checkRequestOpt(opts, subject);
            this._request(subject, stream, req, opts).then((msg)=>{
                clearTimeout(time);
                handler(null, msg, callback);
            }).catch((err)=>{
                clearTimeout(time);
                return Promise.reject(new Error(Util.format('callSync failed, request err:%s', err.message)));
            });
        }
    }

    /**
     * @name: notify
     * @msg: 广播消息不接收处理响应
     * @param {Request} request
     * @return {*}
     */    
    async notify(request) {
        Assert(request, 'call failed, invalid request');
        let req = NRPC.encode(this.encodeType, request, EncodeType1.Request);
        let subj = this.subject + '.' + NRPC.getCallerFuncName();

        try {
            await this.client.publish(subj, req);
            return Promise.resolve();
        } catch (err) {
            return Promise.reject(new Error(Util.format('notify failed, err:%s', err.message)));
        }
    }

    /**
     * @name: _checkRequestOpt
     * @msg: 检查请求配置并修正qos
     * @param {RequestOpt} opt
     * @param {string} subject
     * @return {Promise}
     */    
    async _checkRequestOpt(opt, subject) {
        if (!opt) {
            opt = new RequestOpt();
        }

        Assert(opt.timeout !== 0, '_checkRequestOpt failed, invalid timeout');

        try {
            let stream = await this.client.streamNameBySubject(subject);
            if (opt.getQos() == Qos.Qos0 && opt.getMsgID() != '') {
                opt.setQos(Qos.Qos2);
            } else {
                opt.setQos(Qos.Qos1);
            }
            return Promise.resolve(stream);
        } catch (err) {
            if (opt.getQos() !== Qos.Qos0 ) {
                opt.setQos(Qos.Qos0);
            }
        }

        return Promise.resolve('');
    }

    /**
     * @name: _request
     * @msg: 请求响应消息
     * @param {string} subject
     * @param {string} stream
     * @param {Uint8Array} data
     * @param {RequestOpt} opts
     * @return {Promise}
     */    
    async _request(subject, stream, data, opts) {
        let reply = Nats.createInbox() + '.' + subject;
        console.log(Util.format('request for subject:%s by %s', subject, reply));
        try {
            //订阅接收的响应消息主题
            let sub = await this.client.subscribe(reply, null, null, opts.timeout);
            //发送主题消息
            //TODO JetStreamPublishOptions不存在ackwait选项
            let pubOpts = {
                expect: {
                    streamName: stream,
                }
            };
            if (opts.getQos() === Qos.Qos2 && opts.getMsgID() != '') {
                pubOpts.msgID = opts.getMsgID();
            }

            let msg = {
                subject: subject,
                data: data,
                header: Nats.headers(),
            };
            msg.header.set(ExpectedReplySujectHdr, reply);
            await this.client.publishMsg(msg, pubOpts);
            
            //接收响应
            for await (const m of sub) {
                console.log(Util.format('_request, recv subject:%s, msg:%s', reply, m.subject));
                if (!m.data) {
                    console.log(Util.format('_request, subscribe msg data null, to continue'));
                    continue;
                }
                return Promise.resolve(m);
            }
        } catch (err) {
            return Promise.reject(new Error(Util.format('_request failed, subscribe err:%s', err.message)));
        }
    }
}

module.exports = {
    NRPC,
    Request,
    Response,
    EncodeType,
    NRPCClient,
    RequestOpt,
    Qos,
}